Makefile

```diff
@@ -27,35 +27,43 @@ GENS= \
   $(EVAL_GEN) \
   $(PRIM_PDF_GEN) \
   $(PRIM_TEXT_GEN)
+GENCODE_DIR=tools/gencode
+GENCODE_EXE=gencode.exe
+GENCODE_BIN=$(GENCODE_DIR)/_build/default/$(GENCODE_EXE)
+GENCODE=$(DUNE) exec --root $(GENCODE_DIR) ./$(GENCODE_EXE) --
+INSTDEF=$(GENCODE_DIR)/vminst.ml
 
 .PHONY: all gen install lib uninstall clean
 
 all: gen
 	$(DUNE) build
 	cp _build/install/default/bin/$(TARGET) .
 
+$(INSTDEF): $(INSTDEF_YAML)
+	$(RUBY) $(GENCODE_RB) --ml $< > $@
+
 gen: $(GENS)
 
-$(ATTYPE_GEN): $(INSTDEF) $(GENCODE)
-	$(RUBY) $(GENCODE) --gen-attype $(INSTDEF) > $@
+$(ATTYPE_GEN): $(INSTDEF)
+	$(GENCODE) --gen-attype > $@
```

Comment:
How about generating codes by dune instead of make? For example, dune generates `__attype.gen.ml` if adding the following lines to `src/dune`.

```
(rule
 (targets __attype.gen.ml)
 (deps    (:gen ../tools/gencode/gencode.exe))
 (action  (with-stdout-to %{targets}
           (run %{gen} --gen-attype))))
```

(Ref: [Using a custom code generator](https://jbuilder.readthedocs.io/en/latest/quick-start.html?highlight=preprocess#using-a-custom-code-generator) -- dune’s documentation)

Reply:
this?

>  dune migration and removing ruby related files will be an other PR.

Reply:
Oh I'm sorry I skipped the mention of `dune`.